include/mender-update-module.h

```diff
@@ -0,0 +1,132 @@
+ * @param size Artifact file size
+ * @param data Artifact data
+ * @param index Artifact data index
+ * @param length Artifact data length
+ */
+struct mender_update_download_state_data_s {
+    char    *id;
+    char    *artifact_name;
+    char    *type;
+    cJSON   *meta_data;
+    char    *filename;
+    size_t   size;
+    uint8_t *data;
+    size_t   offset;
+    size_t   length;
+    bool     done;
+};
+
+struct mender_update_install_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_verify_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_commit_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_cleanup_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_rollback_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_rollback_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_failure_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+/* The last member allows to type-cast NULL to mender_update_state_data_t,
+ * otherwise we would have to type-cast NULL to one of the other member types
+ * first. */
+/* Should we just use (void *data) as parameter of the callbacks? */
+typedef union mender_update_state_data_u {
+    struct mender_update_download_state_data_s        *download_state_data;
+    struct mender_update_install_state_data_s         *install_state_data;
+    struct mender_update_reboot_state_data_s          *reboot_state_data;
+    struct mender_update_verify_reboot_state_data_s   *verify_reboot_state_data;
+    struct mender_update_commit_state_data_s          *commit_state_data;
+    struct mender_update_cleanup_state_data_s         *cleanup_state_data;
+    struct mender_update_rollback_state_data_s        *rollback_state_data;
+    struct mender_update_rollback_reboot_state_data_s *rollback_reboot_state_data;
+    struct mender_update_failure_state_data_s         *failure_state_data;
+    void                                              *only_for_nicer_type_casting;
+} mender_update_state_data_t;
+
+typedef mender_err_t (*MenderUpdateStateCallback)(mender_update_state_t state, mender_update_state_data_t callback_data);
+
+typedef struct mender_update_module_s {
+    MenderUpdateStateCallback callbacks[N_MENDER_UPDATE_STATES];
+    const char               *artifact_type;
+    bool                      requires_reboot;
+    bool                      supports_rollback;
+} mender_update_module_t;
```

Comment:
Here and in other places, should we omit the struct tag when using typedef?

```suggestion
typedef struct {
    MenderUpdateStateCallback callbacks[N_MENDER_UPDATE_STATES];
    const char               *artifact_type;
    bool                      requires_reboot;
    bool                      supports_rollback;
} mender_update_module_t;
```

Reply:
Does that even compile? :open_mouth:

Reply:
I haven't tried it, but... yes? From what I remember from C the tag is meant to be used as:

```
struct struct_tag {
...
} ;

struct struct_tag my_struct;
```

And having both the tag and the typedef, will let you do both:
```
struct mender_update_module_s something1;
mender_update_module_t something2;
```

(risky business, I am sending this comment without trying it locally :swimmer: )

Reply:
AFAIK, `struct whatever {...}` creates a new type `struct whatever`. The `typedef` gives it a shorter name. One can surely do `struct {...} whatever;` which declares a variable. I guess `typedef struct {...} whatever;` ignores the fact that the `struct {...}` is an anonymous type. 

I believe it's a good practice to name the `struct` as well, but I don't have a strong opinion here. Anyway, it's a topic for a separate PR changing this in other places too.